btcec/schnorr/signature.go

```diff
@@ -0,0 +1,557 @@
+// Copyright (c) 2013-2017 The btcsuite developers
+// Copyright (c) 2015-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package schnorr
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	ecdsa_schnorr "github.com/decred/dcrd/dcrec/secp256k1/v4/schnorr"
+)
+
+const (
+	// SignatureSize is the size of an encoded Schnorr signature.
+	SignatureSize = 64
+
+	// scalarSize is the size of an encoded big endian scalar.
+	scalarSize = 32
+)
+
+var (
+	tagHashAux = []byte("BIP0340/aux")
+
+	tagHashNonce = []byte("BIP0340/nonce")
+
+	tagHashChallenge = []byte("BIP0340/challenge")
+)
+
+var (
+	// rfc6979ExtraDataV0 is the extra data to feed to RFC6979 when
+	// generating the deterministic nonce for the BIP-340 scheme.  This
+	// ensures the same nonce is not generated for the same message and key
+	// as for other signing algorithms such as ECDSA.
+	//
+	// It is equal to SHA-256([]byte("BIP-340")).
```

Comment:
I couldn't find any documentation for this value, I assume it is up to the implementation to pick it?

Reply:
Check out this section of the BIP: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#alternative-signing 

> This can be most reliably accomplished by not reusing the same private key across different signing schemes. For example, if the rand value was computed as per RFC6979 and the same secret key is used in deterministic ECDSA with RFC6979, the signatures can leak the secret key through nonce reuse.

To avoid that leak, we bind the randomness value derived from RFC6979 to the new signing context. As a result, given the same message, we'll generate a distinct nonce for ECDSA vs Schnorr, thereby alleviating any concerns here re leaking a secret key. 

Note that this signing path is _optional_ with how things are laid out atm. Using RFC 6979 here lets us have signing be deterministic as it is for ECDSA today, and doesn't require the caller to pass in extra randomness themselves `authNonce`.

Reply:
Ah I see, yeah that's a value we've chosen here, inspired by the other prefixes used for the tagged hash implementation across taproot. In theory, we could eventually make _another_ BIP that just specifies the usage of this phrase for those wallets that want to default to deterministic signatures. 

IMO unless you're doing some fancy zkp to have a hardware wallet prove proper nonce generation, then you just want to go with deterministic signatures, as then you rely less on your platform's RNG outside of initial seed generation 9which may have taken place elsewhere).